### ISSUE TITLE
"hinge loss" in in "loss_metric" compare to "standard/general hinge loss"

### ISSUE THREAD

**Issue by RovingBit** [davisking/dlib#1023]
Standard/General hinge loss:
loss = MAX(0, L(P, P+) - L(P, P-)  + C)

Hinge loss in dlib's loss_metric implementation:
loss(postive) = MAX(0, L(P, P+) - dist_thresh + C)
loss(negtive) = MAX(0, dist_thresh - L(P, P-)  + C)
loss = loss(postive) + loss(negtive)

It seems that all "sphere's radius" in dlib's impl have the same "radius < dist_thresh" but the standard hinge loss.

I want to train a network to compare images similarity, both in the same label(johns) and in different labels(john/david/tom...).

In the same label(john1.jpg compare to johnx.jpg):
I think dlib's "hinge loss" implementation will perform better

But how it is performed in different labels(johnx.jpg compare to tomx.jpg):
Which one is better in theory?
Dose dlib's impl of "hinge loss" is a optimized one?
What's the differences?

my english is bad, sorry.
thank you and for your great dlib!

best regards.

**Comment by davisking** [comment]
The regular hinge loss only applies to binary classification problems.  So you can't use it for metric learning, it simply doesn't apply.  You have to do something like what dlib's metric learning tool does.  That's not to say that you can't define metric learning losses that don't have any built in sense of scale, which is maybe what you are asking about.  I wouldn't do that though since something needs to lend a scale to the learning.  People will often use a l2 normalization layer to force a scale at the network output and then use a loss that doesn't account for scale.  I'm not a huge fan of that approach though.

**Comment by RovingBit** [comment]
Thanks for your answer.

But in some deep rank paper and in the FaceNet relatated paper by google, it seems they both use "triplet" and "regular hinge loss" for loss layer.

Do I mis-understand something?
If these questions are too stupid,  I'll close this issue.

Thanks

**Comment by davisking** [comment]
Yes, people do lots of different things.

**Comment by RovingBit** [comment]
After some learn, I found this question is not about hinge loss, but about "**triplet loss function**".

Dlib's triplet loss impl is a optimized version than regular impls.

**Related paper**:
[2] D. Cheng, Y. Gong, S. Zhou, J. Wang, and N. Zheng. Person re-identification by multi-channel parts-based cnn with **improved triplet loss function**. In CVPR, 2016

The key-idea is the newly added restrict-condition "dist_thresh" apllied to every class.
This will results more concentrated in the same class, and the training will converges faster

**In other hand**:
In real world, differences in the same class of verhicles(car/bus/planes) are more bigger than in apples(red apple/green apple/etc..)

But the "dist_thresh" appied to every classes will flatten/erases this between classes.

**So,**
For face identify, this must be a improvement, because we dont care about how similar between apples and bananas, but the red-apple must identified into apples.

**Question**
1. Does I mis-understand?
2. For image similar comparing usage, which one is better? regular "triplet loss function" or dlib's optimized one?

If anyone feel these dicussions is to basic or just boring, I'll re-close this topic.

Thanks.
